src/com/facebook/buck/testrunner/JUnitRunner.java

```diff
@@ -370,6 +370,17 @@ public void testFinished(Description description) throws Exception {
               stdErr.length() == 0 ? null : stdErr.toString()));
     }
 
+    @Override
+    public void testRunFinished(Result runResult) {
```

Comment:
This fixes when Runner start test and then encounters failure that prevents it from continuing the run.

Reply:
Previously `public void testAssumptionFailure(Failure failure) {` would just swallow the error since `resultListener` is not null, but `testFinished` is never called, so this failure is lost.

Reply:
would it not make sense to change line 399 then??
We can check whether the failure is actually an assumptionViolationException, or else record as a failure.

Reply:
that's not necessary `testAssumptionFailure`

Here's how rest runner execution works in the bugged example: 

* call `testStarted `
* call `testFailure ` (this is attached to test, so we don't record it as unbounded, since we record all of these in `testFinished`)
* do not call `testFinished` (since this was kind of failure that prevents runner to continue)
* call `testRunFinished` 

since during call of the method `testFailure` we don't yet know if runner will call `testFinished` or not (normally we'd like to record all the failures in `testFinished`) we can not make a decision yet. 

 `testFinished` wraps up the test and makes `resultListener = null`, so check `resultListener != null` detects where some test was started but not finished. In this case we otherwise will lose the error causing runner to terminate the execution.